dislib/classification/rf/decision_tree.py

```diff
@@ -215,213 +176,245 @@ def compute_split(tree_path, sample, n_features, features_mmap, y_s, n_classes,
+    subtree = Node()
+    tree_traversal = [(subtree, sample, y_s, 0)]
     while tree_traversal:
-        tree_path, sample, y_s, depth = tree_traversal.pop()
+        node, sample, y_s, depth = tree_traversal.pop()
         if depth < max_depth:
-            if use_sklearn_internally and n_features * len(
-                    sample) <= sklearn_max_elements:
+            if use_sklearn and n_features * len(sample) <= sklearn_max:
+                if max_depth == np.inf:
+                    sklearn_max_depth = None
+                else:
+                    sklearn_max_depth = max_depth - depth
                 dt = SklearnDTClassifier(max_features=m_try,
-                                         max_depth=None
-                                         if max_depth == np.inf
-                                         else max_depth - depth)
-                sample, new_indices, sample_weight = \
-                    np.unique(sample, return_index=True, return_counts=True)
+                                         max_depth=sklearn_max_depth)
+                unique = np.unique(sample, return_index=True,
+                                   return_counts=True)
+                sample, new_indices, sample_weight = unique
                 x = get_sample_attributes(samples_file, sample)
                 y_s = y_s[new_indices]
                 dt.fit(x, y_s, sample_weight=sample_weight, check_input=False)
-                node_list_to_persist.append(TreeWrapper(tree_path, dt))
+                node.content = SkTreeWrapper(dt)
             else:
-                node, left_group, y_l, right_group, y_r = compute_split(
-                    tree_path, sample, n_features, features_mmap,
-                    y_s, n_classes, m_try)
-                node_list_to_persist.append(node)
-                if isinstance(node, InternalNode):
-                    tree_traversal.append(
-                        (tree_path + 'R', right_group, y_r, depth + 1))
-                    tree_traversal.append(
-                        (tree_path + 'L', left_group, y_l, depth + 1))
+                split = compute_split(sample, n_features, y_s, n_classes,
+                                      m_try, mmap)
+                node_info, left_group, y_l, right_group, y_r = split
+                node.content = node_info
+                if isinstance(node_info, InnerNodeInfo):
+                    node.left = Node()
+                    node.right = Node()
+                    tree_traversal.append((node.right, right_group, y_r,
+                                           depth + 1))
+                    tree_traversal.append((node.left, left_group, y_l,
+                                           depth + 1))
         else:
-            node_list_to_persist.append(build_leaf(y_s, tree_path))
-    return node_list_to_persist
+            node.content = compute_leaf_info(y_s, n_classes)
+    return subtree
+
+
+@task(returns=list)
+def collect(*object_list):
```

Comment:
What is the purpose of this function?

Reply:
Transforms an array of future objects (tree nodes) to a single future object representing the array, to reduce overhead on the predict phase where this array is used multiple times. This is also the only way to pass 2 arrays of future objects to one task (you can use the star operator only once), but I'm not sure if I do this.
Anyways, I changed the name of the method to `_merge`, which probably fits better.